internal/integration/unified/client_entity.go

```diff
@@ -38,6 +38,52 @@ var securitySensitiveCommands = []string{
 	"createUser", "updateUser", "copydbgetnonce", "copydbsaslstart", "copydb",
 }
 
+// eventSequencer allows for sequence-based event filtering for
+// awaitMinPoolSizeMS support.
+//
+// Per the unified test format spec, when awaitMinPoolSizeMS is specified, any
+// CMAP and SDAM events that occur during connection pool initialization
+// (before minPoolSize is reached) must be ignored. We track this by
+// assigning a monotonically increasing sequence number to each event as it's
+// recorded. After pool initialization completes, we set eventCutoffSeq to the
+// current sequence number. Event accessors for CMAP and SDAM types then
+// filter out any events with sequence <= eventCutoffSeq.
+//
+// Sequencing is thread-safe to support concurrent operations that may generate
+// events (e.g., connection checkouts generating CMAP events).
+type eventSequencer struct {
+	counter atomic.Int64
+	cutoff  atomic.Int64
+
+	mu sync.RWMutex
+
+	// pool events are heterogeneous, so we track their sequence separately
+	poolSeq        []int64
+	seqByEventType map[monitoringEventType][]int64
+}
+
+// setCutoff marks the current sequence as the filtering cutoff point.
+func (es *eventSequencer) setCutoff() {
+	es.cutoff.Store(es.counter.Load())
+}
+
+// recordEvent stores the sequence number for a given event type.
+func (es *eventSequencer) recordEvent(eventType monitoringEventType) {
+	next := es.counter.Add(1)
+
+	es.mu.Lock()
+	es.seqByEventType[eventType] = append(es.seqByEventType[eventType], next)
```

Comment:
Instead of storing every sequence number per event type, can we only store the index when the cutoff happened?

E.g.
```go
func (es *eventSequencer) recordEvent(...) {
	es.mu.Lock()
	defer es.mu.Unlock()

	if !es.isCutoff {
		es.cutoffByEventType[eventType]++
	}
}

func (es *eventSequencer) setCutoff() {
	es.mu.Lock()
	defer es.mu.Unlock()
	es.isCutoff = true
}
```

That would also simplify `filterEventsBySeq` because you just get a subslice index:
```go
func filterEventsBySeq[T any](...) []T {
	// ...
	cutIdx := c.eventSequencer.cutoffByEventType[eventType]
	return localEvents[cutIdx:]
}
```

Reply:
Good idea!

Reply:
While your approach is simpler, it only supports a single cutoff per test. If the unified spec adds scenarios with multiple cutoffs (e.g., awaitMinPoolSizeMS (implicit cutoff) → run X → cutoff at N (either implicit or explicit) → run Y), we’d have to refactor back to the original, more flexible design.